### ISSUE TITLE
Support for defalutValue in @ColumnInfo

### ISSUE THREAD

**Issue by VadimOsovsky** [vitusortner/floor#256]

**Comment by vitusortner** [comment]
Hi! What's your exact use case for this?

**Comment by tulioccalazans** [comment]
@vitusortner In Room, if defaultValue is defined, we don't have to set this property.
This way, in an insertion the field get the default value.

```  
@ColumnInfo(name: "IsActive", defaultValue: "false")
  bool active;
```
It is a useful feature.

**Comment by mqus** [comment]
Why not assign a default value in the class description by changing

> ```dart
> @ColumnInfo(name: "IsActive", defaultValue: "false")
>   bool active;
> ```
to 
```dart
@ColumnInfo(name: "IsActive")
  bool active=false;
```
?
This should cover 90% of the use cases. Could help me understand why that won't work in your case?

**Comment by VadimOsovsky** [comment]
@mqus having non final fields isn't something I usualy do unless really needed

**Comment by mqus** [comment]
Sorry, you had no `final`s in your example code, so I assumed otherwise :wink: 

What about
```dart
  @ColumnInfo(name: "IsActive")
  final bool active;

  YourClass(this.id, this.foo, {this.bar, this.active=false});
```
?
here, `bar` would have the default value `null` and `active` would have the default value `false`.

**Comment by mqus** [comment]
Tbh, If I would implement this, I would probably do it in the same manner as [floor had done it](https://cs.android.com/androidx/platform/frameworks/support/+/androidx-master-dev:room/common/src/main/java/androidx/room/ColumnInfo.java;l=86;drc=4f917f20e63afc653a6092ad87cfbb10bcc1552c), Meaning that the generated `CREATE` query for the entity would set the default value, but it won't ever be used because the `@insert` queries do not omit fields which were set to null. To be able to use that default value, you would have to use queries we still don't support.

**Comment by cassioseffrin** [comment]
Seizing the opportunity of this thread. Could I create a default value as the output a query:
Default value: SELECT strftime('%s', 'now') || (last_insert_rowid()+1)

@ColumnInfo(name: "timestampId", defaultValue: SELECT strftime('%s', 'now') || (last_insert_rowid()+1))

I know It's a little tricky but It's all I need .

May I can do it  making an alter table onCreate event. However if I can do it as a default value will be much more clear.

👍

**Comment by mqus** [comment]
As I wrote before, building a `CREATE TABLE` statement which sets the default with literal values or expressions (like you wrote) is very easy. But selectively omitting fields when creating an `@insert` function is hard[1]! If we don't omit them (like we do now) we set the column of this entity to its value (possibly null) and sqlite won't use the default value, ever.
To use this, you would have to use `@Query('INSERT ...')` dao methods, which in turn don't play nicely with streams (yet). I think it would be cleaner to set default values using dart provisions right now or using triggers, as you wrote.

I think we can accept contributions for the `defaultValue` field in `@ColumnInfo` and its implementation, even if its not usable in `@insert` methods (like in room). What do you think about this, @vitusortner ?

[1] You would have to decide at runtime(slow, for each entity separately) if you want to add single columns to the database or not. Now, the main issue is, when do you omit the columns? when the value is null? What about fields that are nullable even in the database? sqlite makes a difference between inserting `null` values and values that were not provided, a distinction that is impossible with the dart type system (and most others). So if we have a nullable column with a default value, should null be inserted or the default value?
Floor could provide a fixed solution to this edge case or ignore it but this is not as clear-cut as I would like it.

**Comment by vitusortner** [comment]
Contributions for `ColumnInfo.defaultValue` are very welcome, indeed!

**Comment by cassioseffrin** [comment]
Hello,

Any news about  ColumnInfo.defaultValue?

**Comment by leononly** [comment]
Yes this feature is welcome indeed, as we might send the .db file to other system to implement. I hope we can have this feature soon.

**Comment by lucasdidur** [comment]
Any update about this? We are in 2024
